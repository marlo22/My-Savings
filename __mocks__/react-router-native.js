module.exports = {
  useHistory: jest.fn(() => ({
    push: jest.fn(),
    goBack: jest.fn()
  }))
};