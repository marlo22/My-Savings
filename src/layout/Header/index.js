import React from 'react';
import PropTypes from 'prop-types';
import useRouter from 'use-react-router';
import styled from 'styled-components';

import { GlobalContextConsumer } from '../../context/Global';

import { Header as NativeBaseHeader, Left, Right, Body, Button, Icon, Title } from 'native-base';

const BackButton = styled(Button)`
  padding: 0 5px;
`;

const HeaderWrapper = styled.View`
  text-align: center;
`;

const HeaderTitle = styled(Title)`
  margin: 0 5px;
`;

export default function Header({ title }) {
  const { history } = useRouter();

  return (
    <GlobalContextConsumer>
      {context => (
        <NativeBaseHeader>
          <Left>
            <BackButton onPress={history.goBack}>
              <Icon type="AntDesign" name="back" />
            </BackButton>
          </Left>
          <Body>
            <HeaderWrapper>
              <HeaderTitle>
                {title}
              </HeaderTitle>
            </HeaderWrapper>
          </Body>
          <Right>
            <Button onPress={context.toggleNav}>
              <Icon
                type="AntDesign"
                name={context.isNavOpen ? 'menufold' : 'menuunfold'}
              />
            </Button>
          </Right>
        </NativeBaseHeader>
      )}
    </GlobalContextConsumer>
  );
}

Header.propTypes = {
  title: PropTypes.string
};