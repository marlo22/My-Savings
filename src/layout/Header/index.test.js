import React from 'react';
import { useHistory } from 'react-router-native';
import { shallow } from 'enzyme';

import Header from './';

const toggleNav = jest.fn();
const isNavOpen = false;

const useContext = jest
  .spyOn(React, 'useContext')
  .mockImplementation(() => ({
    toggleNav,
    isNavOpen
  }));

const goBack = jest.fn();

useHistory.mockImplementation(() => ({
  goBack
}));

const title = 'Dashboard';
const hasNavMenu = true;

const setup = () => shallow(
  <Header
    title={title}
    hasNavMenu={hasNavMenu}
  />
);

let component = setup();

describe('<Header />', () => {
  test('should render without errors', () => {
    expect(component).toBeTruthy();
  });

  test('snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  test('click back button should call go back router\'s function', () => {
    const backBtn = component.find('Styled(Styled(Button))');
    backBtn.simulate('press');
    expect(goBack).toHaveBeenCalled();
  });

  test('should render passed header title', () => {
    const header = component.find('Styled(Styled(Title))');
    expect(header.text()).toBe(title);
  });

  test('click navigation button should call toggleNav context\'s function', () => {
    const navBtn = component.find('Styled(Button)');
    navBtn.simulate('press');
    expect(toggleNav).toHaveBeenCalled();
  });

  test('if nav menu is close should render "menuunfold" icon', () => {
    const menuIcon = component.find('Styled(Icon)').last();
    expect(menuIcon.prop('name')).toBe('menuunfold');
  });

  test('if nav menu is open should render "menufold" icon', () => {
    useContext.mockImplementation(() => ({
      toggleNav,
      isNavOpen: true
    }));

    component = setup();

    const menuIcon = component.find('Styled(Icon)').last();
    expect(menuIcon.prop('name')).toBe('menufold');
  });

  test('if hasNavMenu prop is falsy should not render toggle navigation button', () => {
    component.setProps({ hasNavMenu: false });

    const navBtn = component.find('Styled(Button)');    
    expect(navBtn).toHaveLength(0);
  });
});