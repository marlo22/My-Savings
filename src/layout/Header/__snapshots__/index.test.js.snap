// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> snapshot 1`] = `
<Styled(Header)>
  <Styled(Left)>
    <Styled(Styled(Button))
      onPress={[MockFunction]}
    >
      <Styled(Icon)
        name="back"
        type="AntDesign"
      />
    </Styled(Styled(Button))>
  </Styled(Left)>
  <Styled(Body)>
    <Styled(View)>
      <Styled(Styled(Title))>
        Dashboard
      </Styled(Styled(Title))>
    </Styled(View)>
  </Styled(Body)>
  <Styled(Right)>
    <Styled(Button)
      onPress={[MockFunction]}
    >
      <Styled(Icon)
        name="menuunfold"
        type="AntDesign"
      />
    </Styled(Button)>
  </Styled(Right)>
</Styled(Header)>
`;
